gtfs/spec/en/reference.md

```diff
@@ -360,7 +360,7 @@ To process the cost of a multi-leg journey:
 | `duration_limit` | Non-negative integer | Optional | Defines the duration limit of the transfer.<br><br>Must be expressed in integer increments of seconds.<br><br>If there is no duration limit, `fare_transfer_rules.duration_limit` must be empty. |
 | `duration_limit_type` | Enum | **Conditionally Required** | Defines the relative start and end of `fare_transfer_rules.duration_limit`.<br><br>Valid options are:<br>`0` - Between the departure fare validation of the first leg and the arrival fare validation of the last leg.<br>`1` - Between the departure fare validation of the first leg and the departure fare validation of the last leg.<br>`2` - Between the arrival fare validation of the first leg and the departure fare validation of the last leg.<br>`3` - Between the arrival fare validation of the first leg and the arrival fare validation of the last leg.<br><br>Conditionally Required:<br>- **Required** if `fare_transfer_rules.duration_limit` is defined.<br>- **Forbidden** if `fare_transfer_rules.duration_limit` is empty. |
-| `fare_transfer_type` | Enum | **Conditionally Required** | Indicates the cost processing method of leg-to-leg travel (i.e., sub-journey).<br><br>Valid options are:<br>`0` - The cost of the sub-journey is the total calculated cost of the preceding leg(s) plus the cost of `fare_transfer_rules.amount`.<br>`1` - The cost of the sub-journey is the total calculated cost of the preceding leg(s) plus the cost of `fare_transfer_rules.amount` plus the cost of `fare_leg_rules.amount` for `fare_transfer_rules.to_leg_group_id`.<br>`2` - The cost of the sub-journey is the cost of the leg with the most expensive `fare_leg_rules.amount` plus the cost in `fare_transfer_rules.amount`.<br>`3` - The cost of the two-leg sub-journey is the cost in `fare_transfer_rules.amount`.<br><br>Conditionally Required:<br>- **Required** if `fare_transfer_rules.amount` is defined.<br>- **Forbidden** otherwise. |
+| `fare_transfer_type` | Enum | **Conditionally Required** | Indicates the cost processing method of a sub-journey.<br><br>Valid options are:<br>`0` - The cost of the sub-journey is the total calculated cost of the preceding leg(s) and transfer(s) plus the cost of `fare_transfer_rules.amount`.<br>`1` - The cost of the sub-journey is the total calculated cost of the preceding leg(s) and transfer(s) plus the cost of `fare_transfer_rules.amount` plus the cost of `fare_leg_rules.amount` for `fare_transfer_rules.to_leg_group_id`.<br>`2` - The cost of the sub-journey is the cost of the leg with the most expensive `fare_leg_rules.amount` plus the cost in `fare_transfer_rules.amount`.<br>`3` - The cost of the sub-journey from the preceding leg(s) and transfer(s) to the transferred leg is the cost in `fare_transfer_rules.amount`.<br><br>Conditionally Required:<br>- **Required** if `fare_transfer_rules.amount` is defined.<br>- **Forbidden** otherwise. |
```

Comment:
Either there's a typo or I don't understand #3: "the cost of a sub-journey FROM the preceding leg(s) and transfer(s) TO the transferred leg IS the cost in fare_transfer_rules.amount" Shouldn't this rather be: "the cost of a sub-journey IS the cost in fare_transfer_rules.amount"? I mean it's easy to see what's the difference between #0 and #1, isn't #3 similar to that difference? I mean: #0: previous SUM + transfer.amount, #1: previous SUM + transfer.amount + leg.amount, #3: only transfer.amount (aka drop: the previous amount)? If this is the case then maybe it would also be easier if we ordered them this way:
0: previous SUM + transfer.amount + leg.amount
1: previous SUM + transfer.amount
2: transfer.amount
3: most expensive leg + transfer.amount

And looking at these I also have another questions:
- where is "previous SUM + leg.amount"? (or is it "previous SUM + transfer.amount + leg.amount" with transfer.amount=0?)
- but then isn't "previous SUM + transfer.amount" == "previous SUM + transfer.amount + leg.amount" with leg.amount=0?

And BTW there are also a real world use cases like: the cost of the sub-journey is calculated by the "furthest transit stop from the origin" (Sydney)

Reply:
Please see my comment and illustrations [here](https://github.com/google/transit/pull/286#discussion_r759636579).

Reply:
> Shouldn't this rather be: "the cost of a sub-journey IS the cost in fare_transfer_rules.amount"? 

This is a cleaner description (I like it) but it doesn't capture what happens when we "chain" multiple legs in a journey (i.e., the interaction of `fare_transfer_type` illustrated [here](https://github.com/google/transit/pull/286#discussion_r759636579)).

I'm wondering if in light of the brainstorming on `fare_transfer_type` interactions, we should drop the concept of "sub-journey" all-together in these descriptions. As you have pointed out, it's not clear what a sub-journey is, and the fact that it can actually be any contiguous subset of a journey is not actually helpful when describing `fare_transfer_type` interactions and their chaining.

Here is my proposal to clean up the description of `fare_transfer_type`:

> Indicates the cost processing method of transferring between legs in a journey:
> 
> ![](https://user-images.githubusercontent.com/62957380/144116057-d5cf7ef0-1a67-4e65-922c-10140e410459.png) 
> Valid options are:
> `0` - `fare_leg_rules.amount` plus the cost of `fare_transfer_rules.amount`; A + AB.
> `1` - `fare_leg_rules.amount` plus the cost of `fare_transfer_rules.amount` plus the cost of `fare_leg_rules.amount` for the transferred leg; A + AB + B.
> `2` - The leg with the most expensive `fare_leg_rules.amount` plus the cost of `fare_transfer_rules.amount` ; MAX(A, B) + AB.
> `3` - `fare_transfer_rules.amount`; AB.
> 
> Cost processing interactions of transferring between legs in a journey with more than 2 legs:
>
> ![](https://user-images.githubusercontent.com/62957380/144116095-8a0aedae-5235-4782-a58f-e0b3644dd5f2.png)
> | `fare_transfer_type` | Processing A > B | Processing B > C |
> | ----- | ----- | ----- | 
> | `0` | A + AB | S + BC |
> | `1` | A + AB + B | S + BC + C |
> | `2` | MAX(A, B) + AB | S + MAX(B, C) + BC |
> | `3` | AB | S + BC |
> 
> Where S indicates the total calculated cost of the preceding leg(s) and transfer(s).

Does this clarify all possible uses?

Reply:
I've gone ahead and applied this patch. Let me know if it looks good to you.

Reply:
> > Shouldn't this rather be: "the cost of a sub-journey IS the cost in fare_transfer_rules.amount"?
> 
> This is a cleaner description (I like it) but it doesn't capture what happens when we "chain" multiple legs in a journey (i.e., the interaction of `fare_transfer_type` illustrated [here](https://github.com/google/transit/pull/286#discussion_r759636579)).

Still the 1st part of the sentence is confusing IMHO. What about something different: "The cost of the sub-journey is [only?] the cost of fare_transfer_rules.amount which overrides the costs from the previous legs and transfers"?

> 
> I'm wondering if in light of the brainstorming on `fare_transfer_type` interactions, we should drop the concept of "sub-journey" all-together in these descriptions. As you have pointed out, it's not clear what a sub-journey is, and the fact that it can actually be any contiguous subset of a journey is not actually helpful when describing `fare_transfer_type` interactions and their chaining.
> 
> Here is my proposal to clean up the description of `fare_transfer_type`:
> 
> > Indicates the cost processing method of transferring between legs in a journey:
> > ![](https://user-images.githubusercontent.com/62957380/144116057-d5cf7ef0-1a67-4e65-922c-10140e410459.png)
> > Valid options are:
> > `0` - `fare_leg_rules.amount` plus the cost of `fare_transfer_rules.amount`; A + AB.
> > `1` - `fare_leg_rules.amount` plus the cost of `fare_transfer_rules.amount` plus the cost of `fare_leg_rules.amount` for the transferred leg; A + AB + B.
> > `2` - The leg with the most expensive `fare_leg_rules.amount` plus the cost of `fare_transfer_rules.amount` ; MAX(A, B) + AB.
> > `3` - `fare_transfer_rules.amount`; AB.
> > Cost processing interactions of transferring between legs in a journey with more than 2 legs:
> > ![](https://user-images.githubusercontent.com/62957380/144116095-8a0aedae-5235-4782-a58f-e0b3644dd5f2.png)
> > 
> > 
> > 
> > `fare_transfer_type`
> > Processing A > B
> > Processing B > C
> > 
> > 
> > 
> > 
> > `0`
> > A + AB
> > S + BC
> > 
> > 
> > `1`
> > A + AB + B
> > S + BC + C
> > 
> > 
> > `2`
> > MAX(A, B) + AB
> > S + MAX(B, C) + BC
> > 

I think this contradicts the meaning of the sentence and should instead be: MAX(A, ..., C) + BC (or something similar, the main point is that "S+" shouldn't be there) but it also shows that there is some kind of problem with this rule (even now, without mentioning "sub-journeys"). For example it's easy to imagine the following situation: on buses you pay per distance or whatever way, on subway legs you only pay for the longest (most expensive) leg. The problem is that when you mix them i.e: Subway1, Bus2, Subway3, Bus4 and the cost would be: max(S1,S3) + B2 + B4, then this notation won't work.

> > 
> > `3`
> > AB
> > S + BC
> > 
> > 
> > 
> > Where S indicates the total calculated cost of the preceding leg(s) and transfer(s).
> 
> Does this clarify all possible uses?

I think we should still think about the non-consecutive cases. In the table for the "more than 2 legs" journeys it is even now implicitly required to have consecutive legs (because of A, B. C and S). I think there should be a way to easily understand from the spec that cases like the above example with the buses and subways having different fare system so the total cost of the full journey will be the sum of the "sub-journeys" (buses + subways) are possible and legal. But of course there could be many more examples (different agencies have different fare system, so the "sub-journeys" in this case would be agency_set1 + agency_set2 + ...)